Show health-check data in dashboard

drewnoakes opened the pull request:
## Description

Fixes #5569

This change plumbs through data about health-checks to the dashboard.

Values are displayed in the resource detail view, in their own accordian section, in a new three-column grid:

<img width="600" src="https://github.com/user-attachments/assets/7e010d82-d36b-4e0e-8d31-e1a77da98fbf">
<img width="600" src="https://github.com/user-attachments/assets/8e75dd23-6934-4d8a-8461-d1e31b5875fd">
<img width="600" src="https://github.com/user-attachments/assets/db6b2bd2-d7a1-4364-986d-d79082e7b50d">

The icon and popup used in the structured log message column to show exceptions was pulled into a new `<ExceptionDetails>` component and reused for health check error data.

![image](https://github.com/user-attachments/assets/d2fbe898-a589-4305-a2ad-1f7c4447324c)

Tooltip handling on the resource grid's state column cells was previously inconsistent. We would sometimes add a tooltip to the icon and text, and other times to the cell's entire bounds. Formatting logic for the resource grid's state column is moved into code-behind, simplifying the razor and ensuring consistent treatment. Tooltips are now always at the cell level, making them easier to hit with the mouse, and making the accessible text consistent.

![image](https://github.com/user-attachments/assets/8d0ce51e-56d7-4345-a231-26f1fc047ed7)

As part of this change, a new `ContentBeforeValue` parameter was added to `GridValue`, as a complement to `ContentAfterValue`.

A new "HealthChecksSandbox" playground was added to make testing of health check data easier.

Resources without health checks appear immediately healthy.

![health-state](https://github.com/user-attachments/assets/94eb5799-bb77-4f4e-bae7-2626fc67aad0)

A placeholder message is shown in the resource details until the first health data arrives:

![image](https://github.com/user-attachments/assets/40639829-4267-4fd7-9d57-e033847a4afe)


## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6069)

JamesNK commented:
![image](https://github.com/user-attachments/assets/49a9d15b-c465-4dd9-8f7a-6721c6769a99)

It would be much easier to tell at a glance the health status by having an icon here to go with the text.

Also, I think the description/exception should be in a popup. That avoids overloading the column with too much information.

You probably don't want to use the PropertyGrid control because of the custom content, but having a custom data grid in a details control is simple enough: https://github.com/dotnet/aspire/blob/d0d3ce55c07baed98891a6e95e294ae6348c347e/src/Aspire.Dashboard/Components/Controls/SpanDetails.razor#L94-L113

JamesNK commented:
![image](https://github.com/user-attachments/assets/3cfc278a-b218-426a-85ab-b91a71379c18)

What happens if wait for is depending on something with multiple instances? Are multiple rows displayed? Also, is the resource name here the full name, e.g. `message-abcxyz` but resolved to message, or the display name, e.g. `message`?

The dashboard tends to use the full name everywhere but it's not obvious because it shortens it when possible. It should keep doing that if possible. But I don't know if the full name is known at this point.

davidfowl commented:
Lots of new test falures.

davidfowl commented on src/Aspire.Hosting/Dashboard/ResourceSnapshot.cs:
```diff
@@ -26,8 +25,9 @@ internal abstract class ResourceSnapshot
     public required ImmutableArray<VolumeSnapshot> Volumes { get; init; }
     public required ImmutableArray<UrlSnapshot> Urls { get; init; }
-    public required HealthStateKind? HealthState { get; set; }
+    public required ImmutableArray<HealthReportSnapshot> HealthReports { get; init; }
```
We want the aggregate state top level as well as exposing individual heath states.

davidfowl commented on src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs:
```diff
@@ -71,12 +74,25 @@ public sealed record CustomResourceSnapshot
+    /// <remarks>
+    /// This property's value is derived from <see cref="HealthReports"/> and is provided for convenience.
+    /// </remarks>
+    public HealthStatus HealthStatus => HealthReports.MinBy(r => r.Status)?.Status ?? HealthStatus.Healthy;
```
This should be a cheap check. Let's just store the aggregate and not make this computed.

davidfowl commented on src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs:
```diff
@@ -53,7 +55,23 @@ public sealed record CustomResourceSnapshot
+    /// <remarks>
+    /// This property's value is derived from <see cref="HealthReports"/> and is provided for convenience.
+    /// </remarks>
+    public HealthStatus HealthStatus { get; private set; } = HealthStatus.Healthy;
```
Null by default?

davidfowl commented on src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs:
```diff
@@ -53,7 +55,23 @@ public sealed record CustomResourceSnapshot
+        init
+        {
+            _healthReports = value;
+            HealthStatus = value.MinBy(r => r.Status)?.Status ?? HealthStatus.Healthy;
```
Why compute it at all? The code that sets this has the precomputed value?

davidfowl commented:
> I think we should have an icon here. It would be much easier to tell at a glance the health status by having an icon here to go with the text.

+ 100, spoke to @drewnoakes about this too.

> Also, I think the description/exception should be in a popup. The structured logs grid does that with exception information. There isn't enough room for 3 columns with text, so a popup avoids overloading the column with too much information.

The popup we have for the exception message today is very annoying as it goes away on hover. Maybe we can make them be the same and then fix both 😄

drewnoakes committed:
Show health-check data in dashboard

This change plumbs through data about health-checks to the dashboard.

Values are displayed in the resource detail view, in their own accordian section, in a new three-column grid.

The icon and popup used in the structured log message column to show exceptions was pulled into a new `<ExceptionDetails>` component and reused for health check error data.

Tooltip handling on the resource grid's state column cells was previously inconsistent. We would sometimes add a tooltip to the icon and text, and other times to the cell's entire bounds. Formatting logic for the resource grid's state column is moved into code-behind, simplifying the razor and ensuring consistent treatment. Tooltips are now always at the cell level, making them easier to hit with the mouse, and making the accessible text consistent.

As part of this change, a new `ContentBeforeValue` parameter was added to `GridValue`, as a complement to `ContentAfterValue`.

A new "HealthChecksSandbox" playground was added to make testing of health check data easier.

drewnoakes commented:
@davidfowl @JamesNK see the updated PR description for info on changes.

- Removed "waits for" info
- Updated presentation of health checks (screenshots in description)

JamesNK commented on src/Aspire.Dashboard/Resources/ControlsStrings.resx:
```diff
@@ -299,6 +305,9 @@
     <value>Volumes</value>
   </data>
+  <data name="ResourceHealthChecksHeader" xml:space="preserve">
+    <value>Health Checks</value>
```
Change to `Health checks`. Other section titles capitalize the first word only.

JamesNK commented on src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor:
```diff
@@ -108,6 +111,60 @@
+                                   HighlightText="@_filter" />
+                    </AspireTemplateColumn>
+                    <AspireTemplateColumn Title="@ControlStringsLoc[nameof(ControlsStrings.StateColumnHeader)]" Class="stateColumn">
+                        <GridValue ValueDescription="@ControlStringsLoc[nameof(ControlsStrings.StateColumnHeader)]"
```
I noticed when I searched for the state name, e.g. `Healthy`, the healthcheck disappeared. I think either don't highlight state when searching, or have the state be detected by searching.

JamesNK commented on src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor:
```diff
@@ -108,6 +111,60 @@
+                                        HealthStatus.Healthy => ((Icon)new Icons.Filled.Size16.Heart(), Color.Success),
+                                        HealthStatus.Degraded => (new Icons.Filled.Size16.HeartBroken(), Color.Warning),
+                                        HealthStatus.Unhealthy => (new Icons.Filled.Size16.HeartBroken(), Color.Error),
+                                        _ => (new Icons.Regular.Size16.Heart(), Color.Success)
```
Could health status be something else?

I think either throw an error (which would break the whole UI so probably not a good idea) or have an unknown icon. Could reuse what the resources grid shows which I think is a grey circle.

drewnoakes replied:
We use a dashed circle for unknown resource status. There's no dashed heart icon, so I'll use the same dashed circle (fluent calls it a hint).

JamesNK commented on src/Aspire.Dashboard/Model/ResourceViewModel.cs:
```diff
@@ -252,3 +252,51 @@ public bool MatchesFilter(string filter) =>
+
+    private static string GetCompositeValue(HealthStatus HealthStatus, string? Description, string? Exception)
+    {
+        StringBuilder builder = new();
```
var

JamesNK commented on src/Aspire.Dashboard/Model/ResourceViewModel.cs:
```diff
@@ -252,3 +252,51 @@ public bool MatchesFilter(string filter) =>
+    {
+        StringBuilder builder = new();
+
+        builder.Append(HealthStatus.ToString());
```
Other places call humanize on healthstatus. Should be consistent.

JamesNK commented on src/Aspire.Dashboard/Model/ResourceViewModel.cs:
```diff
@@ -252,3 +252,51 @@ public bool MatchesFilter(string filter) =>
+        return
+            Name?.Contains(filter, StringComparison.CurrentCultureIgnoreCase) == true ||
+            Description?.Contains(filter, StringComparison.CurrentCultureIgnoreCase) == true ||
+            Exception?.Contains(filter, StringComparison.CurrentCultureIgnoreCase) == true;
```
I think the problem with including exception here is it isn't visible on the page. While searching you don't see text it being highlighted. It could be confusing why the health check is still visible. IMO remove it.

Also, add a check on the humanized result of the healthstatus enum (see earlier comment about it being filtered out). Maybe `HealthStatus.Humanize()` should be cached somewhere. Don't want to repeatedly process it in a Where statement. I don't know how expensive it is.

drewnoakes replied:
Looking at the implementation, it's moderately expensive.

JamesNK commented on src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor:
```diff
@@ -108,6 +111,60 @@
+                        </GridValue>
+                    </AspireTemplateColumn>
+                    <AspireTemplateColumn Title="@ControlStringsLoc[nameof(ControlsStrings.DetailsColumnHeader)]" Class="detailsColumn">
+                        <GridValue ValueDescription="@ControlStringsLoc[nameof(ControlsStrings.DetailsColumnHeader)]"
```
What do you think about having a placeholder value if there aren't details?

![image](https://github.com/user-attachments/assets/d77a0a02-5ccc-4871-a657-4c5deb5bd024)

Something like `(None)` or `(Empty)` might feel better than an empty column.

We don't do that for other values on these grids yet so probably shouldn't start in this PR.

drewnoakes replied:
We're inconsistent on empty values. Even in the resource grid, we have show "(None)" for some columns (like endpoints) but not others (like source).

JamesNK commented on src/Aspire.Dashboard/ResourceService/Partials.cs:
```diff
@@ -48,19 +48,31 @@ public ResourceViewModel ToViewModel(BrowserTimeProvider timeProvider, IKnownPro
+            {
+                return healthStatus switch
+                {
+                    HealthStatus.Unknown or HealthStatus.Healthy => Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus.Healthy,
```
Why is unknown state considered healthy? Intentional or mistake?

If it is intentional, could you add a comment saying why.

drewnoakes replied:
Innocent until proven guilty. Will add a comment.

JamesNK replied:
What does wait for do when a dependency has a health check that returns unknown? This should be consistent with it @mitchdenny 

It would be odd to have a green tick with an unknown status in the UI, but wait for is still waiting for it to enter a health state.

JamesNK replied:
I looked into this more, and we're using the extensions health check enum on the server and the client (dashboard) with an unknown status inbetween on the proto layer.

It's impossible for the app host to return unknown, but a custom resource service implementation could return it.

I think the dashboard should stop using `Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus` and use the enum that reflects all the values that can be returned from the server. That way we don't have to guess what is meant by unknown.

JamesNK replied:
Alternatively, maybe there shouldn’t be an unknown option in the service layer?

drewnoakes replied:
I believe the most recent push addresses the concerns here.

drewnoakes replied:
See 68578f85006decbd721df7d2017a18cf9489ff99 and 14d55aca335107abbe937a52ec2f14921084ec68.

JamesNK commented on src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs:
```diff
@@ -159,6 +172,15 @@ internal void Deconstruct(out string name, out object? value, out bool isSensiti
+/// <param name="Status">The state of the resource, according to the report.</param>
+/// <param name="Description">An optional description of the report, for display.</param>
+/// <param name="Exception">An optional string containing exception details.</param>
+public sealed record HealthReportSnapshot(string Name, HealthStatus Status, string? Description, string? Exception);
```
`Exception` -> `ExceptionDetails`? It is surprising to me to have a property called Exception that is a string and not an `System.Exception`.

drewnoakes replied:
Agree. I've used `ExceptionText` elsewhere so will be consistent with that.

JamesNK replied:
This is partially done. Code in the app host like this snapshot is still `Exception`.

JamesNK commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -191,22 +209,17 @@ message Resource {
+    // The aggregate health state of the resource. Generally reflects data from health_reports, but may differ.
+    optional HealthStatus health_status = 16;
+    // Reports from health checks, about this resource.
+    repeated HealthReport health_reports = 19;
 
     // The resource start time.
     optional google.protobuf.Timestamp started_at = 17;
     // The resource stop time.
     optional google.protobuf.Timestamp stopped_at = 18;
```
I believe these are all new in 9.0 so you can adjust the field numbers to not be out of order.

drewnoakes replied:
Will do. FYI @snehapar9, in case you've already implemented anything in ACA for this proto.

drewnoakes committed:
Merge branch 'main' into drnoakes/health-checks-and-wait-for-details

drewnoakes committed:
Review feedback

JamesNK reviewed (commented):
A couple more things

JamesNK commented:
It's possible to have an exception with no details:

![image](https://github.com/user-attachments/assets/672f038b-d32a-4997-9b17-bb3fb4ad1604)

This looks a little odd. How about in this situation (there are exception details but no text details) default the details text to "Exception details"?

mitchdenny commented on src/Aspire.Hosting/Dashboard/DashboardServiceData.cs:
```diff
@@ -50,13 +48,7 @@ static GenericResourceSnapshot CreateResourceSnapshot(IResource resource, string
                     ExitCode = snapshot.ExitCode,
                     State = snapshot.State?.Text,
                     StateStyle = snapshot.State?.Style,
-                    HealthState = resource.TryGetAnnotationsIncludingAncestorsOfType<HealthCheckAnnotation>(out _) ? snapshot.HealthStatus switch
```
As discussed in our call, this is likely the cause of the incorrect state issues we are seeing in the dashboard. Thinking about it a bit more, its possibly also the reason why we were polling at increased intervals as well (not 100% sure on that though).

Just a reminder to fix this before merging.

davidfowl replied:
What did you see?

drewnoakes replied:
When the resource is created, it has no health reports so appears as "Running". Then the reports arrive, and it transitions to "Running (Unhealthy)", then finally back to "Running".

I'm going to ensure that a resource with a health check annotation that has no reports is still considered unhealthy.

There was an additional behaviour we saw where the health check polling was happening more frequently than we expected it to. I'll fix this and we can check whether things return to what we expected to see.

davidfowl replied:
Don't have any health reports should not be the same as the health reports have not arrived yet. So treat null different to 0.

drewnoakes replied:
This has been fixed in the latest push. Resources that expect health checks have a null health state, which the UI shows as unhealthy.

Resources without health checks appear immediately healthy.

I've also added a placeholder message in the resource details area that appears until the first health data arrives.

See the PR description for new video/screenshots.

drewnoakes commented on src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs:
```diff
@@ -45,6 +45,11 @@ public partial class ResourceDetails
             .AsQueryable();
 
+    private IQueryable<HealthReportViewModel> FilteredHealthReports =>
+        Resource.HealthReports
```
Ordered when we create `ResourceViewModel` from the protobuf so we don't order twice per render.

drewnoakes committed:
Review feedback

Mrxx99 commented:
Was the displaying of WaitFor removed from this PR because
a) It will be done in a separate PR
b) The ideas was scrapped?

I hope it's a) because it seems quite useful, esp. if it would show the wait status as icon (either waiting/not waiting or the status icon of the resource which it is waiting on would be a good solution)

davidfowl commented:
Yes it was removed, we want to design a relationships feature and waiting will be one kind of relationship.

drewnoakes committed:
Merge branch 'main' into drnoakes/health-checks-and-wait-for-details

drewnoakes committed:
Track when expected health check absent

drewnoakes committed:
HealthReportViewModel is no longer IPropertyGridItem

drewnoakes committed:
Show first line of exception when no description exists

drewnoakes committed:
Show message when waiting for expected health data

drewnoakes commented:
> It's possible to have an exception with no details:
> 
> ![image](https://private-user-images.githubusercontent.com/303201/373462383-672f038b-d32a-4997-9b17-bb3fb4ad1604.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgwNDM3ODIsIm5iZiI6MTcyODA0MzQ4MiwicGF0aCI6Ii8zMDMyMDEvMzczNDYyMzgzLTY3MmYwMzhiLWQzMmEtNDk5Ny05YjE3LWJiM2ZiNGFkMTYwNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwNFQxMjA0NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YTQ0NjUzMTZhZWY0YjgyYjdlMTEzNjJiOGI2OTFkMzNjNDRkMTUxNGIzNTNlNGJlMTNhNjcxNDZkODBmMzkyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.wk5MtbRin78Dda9_HNqYgmtigTQ2ozWbeujEYBiuVgk)
> 
> This looks a little odd. How about in this situation (there are exception details but no text details) default the details text to "Exception details" or "Error thrown", etc.

I updated the column to show the first line of the exception text when an exception exists. The icon won't look lonely any more.

![image](https://github.com/user-attachments/assets/2b8e112d-6818-4435-8795-246c779af609)

drewnoakes commented on src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs:
```diff
@@ -48,9 +48,26 @@ public sealed record CustomResourceSnapshot
+    /// The health reports for this resource.
+    /// </summary>
+    /// <remarks>
+    /// May be zero or more. If there are no health reports, the resource is considered healthy.
```
```suggestion
    /// May be zero or more. If there are no health reports, the resource is considered healthy
    /// so long as no heath checks are registered for the resource.
```

drewnoakes committed:
Improve doc comment

JamesNK reviewed (approved):
The only outstanding thing is guessing a status for unknown. It is still happening. I think the simplest fix is remove UNKOWN from the enum. It's new in 9 so it isn't a breaking change.

There are also a bunch of parameters with init instead of set.

Otherwise, it looks good to me.

drewnoakes commented:
> The only outstanding thing is guessing a status for unknown. It is still happening. I think the simplest fix is remove UNKOWN from the enum. It's new in 9 so it isn't a breaking change.

@JamesNK the item having a zero value seems important, as an omitted field would get that default. I can see a default of both healthy or unhealthy making sense, and having an UNKNOWN item avoids that decision. Do you have a view on that?

JamesNK commented:
This is a best practice guide about having a zero value on an enum: https://protobuf.dev/programming-guides/dos-donts/#unspecified-enum

Personally I think we should rely on implementors to always set a value, and if they don't then they'll observe that not setting a value will default to Healthy. We're not adding this enum to an existing message so we don't need to keep compatibility with someone not setting it because they have an old proto spec.

Or we could follow the guide and have a zero value option that is `UNSET`/`UNSPECIFIED`/`UNDEFINED`/`UNKNOWN`. But the guide says not to assign semantic meaning to it having the unspecified value which I agree with. Either we display a health check in the UI with a status of unknown (ok), or the dashboard throws an error (bad, because it breaks everything else). But I don't like guessing a real status from a missing status from the server.

drewnoakes committed:
Remove unknown health status

drewnoakes committed:
Parameters are mutable

drewnoakes committed:
Fix workload test

drewnoakes committed:
Increase playwright test timeout to 30 sec

It timed out on CI at 15.

drewnoakes commented:
CI failure is a failing test in `MySqlFunctionalTests.VerifyWaitForOnMySqlBlocksDependentResources` which seems unrelated to this PR.

davidfowl committed:
Merge branch 'main' into drnoakes/health-checks-and-wait-for-details
